### ISSUE TITLE
HAProxy fails to reload when local peer name starts with "-x"

### ISSUE THREAD

**Issue by johnsom** [haproxy/haproxy#644]
## Output of `haproxy -vv` and `uname -a`

```
$ haproxy -vv
HA-Proxy version 2.1.4 2020/04/02 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.4.html
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_REGPARM=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 1.1.1d FIPS  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.5
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[TRACE] trace
	[COMP] compression

$ uname -a
Linux workstation 5.6.12-300.fc32.x86_64 #1 SMP Mon May 11 16:47:13 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
AKA: Fedora 32 - haproxy-2.1.4-1.fc32.x86_64
```
**Also seen on:**
```
# haproxy -vv
HA-Proxy version 1.8.8-1ubuntu0.10 2020/04/03
Copyright 2000-2018 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -g -O2 -fdebug-prefix-map=/build/haproxy-RAYurj/haproxy-1.8.8=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2
  OPTIONS = USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE=1 USE_PCRE_JIT=1 USE_NS=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
# uname -a
Linux amphora-86423640-ea6a-4aae-a079-42e06cd4be56 4.15.0-99-generic #100-Ubuntu SMP Wed Apr 22 20:32:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

AKA: Ubuntu 18.04.4 LTS - haproxy 1.8.8-1ubuntu0.10
```
## What's the configuration?

```
Stock configuration as shipped. The configuration does not impact this bug.
```

## Steps to reproduce the behavior

1. Modify you systemd unit file to include a "-L set local peer name" setting that includes "-x" as the prefix of the peer name. Example: "ExecStart=/usr/sbin/haproxy -Ws -f $CONFIG -p $PIDFILE -L -xfoo"
2. sudo systemctl daemon-reload
3. sudo systemctl start haproxy
4. sudo systemctl reload haproxy

## Actual behavior

HAProxy will fail to reload and will print a usage statement.

```
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 systemd[1]: Reloading HAProxy Load Balancer.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]: [WARNING] 140/204836 (2411) : Reexecuting Master process
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]: Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -v displays version ; -vv shows known build options.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -d enters debug mode ; -db only disables background mode.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dM[<byte>] poisons memory with <byte> (defaults to 0x50)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -V enters verbose mode (disables quiet mode)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -D goes daemon ; -C changes to <dir> before loading files.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -W master-worker mode.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -Ws master-worker mode with systemd notify support.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -q quiet mode : don't display messages
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -c check mode : only check config files and exit
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -n sets the maximum total # of connections (2000)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -m limits the usable amount of memory (in MB)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -N sets the default, per-proxy maximum # of connections (2000)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -L set local peer name (default to hostname)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -p writes pids of all children to this file
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -de disables epoll() usage even when available
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dp disables poll() usage even when available
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dS disables splice usage (broken on old kernels)
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dG disables getaddrinfo() usage
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dR disables SO_REUSEPORT usage
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dr ignores server address resolution failures
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -dV disables SSL verify on servers side
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -sf/-st [pid ]* finishes/terminates old pids.
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]:         -x <unix_socket> get listening sockets from a unix socket
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]: HA-Proxy version 1.8.8-1ubuntu0.10 2020/04/03
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2411]: Copyright 2000-2018 Willy Tarreau <willy@haproxy.org>
May 20 20:48:40 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 systemd[1]: haproxy.service: Main process exited, code=exited, status=1/FAILURE

```

## Expected behavior

We would expect HAProxy to reload successfully like it does with other prefixes.

```
May 20 21:05:08 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 systemd[1]: Reloading HAProxy Load Balancer.
May 20 21:05:08 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2454]: [WARNING] 140/210458 (2454) : Reexecuting Master process
May 20 21:05:08 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 systemd[1]: Reloaded HAProxy Load Balancer.
May 20 21:05:08 amphora-86423640-ea6a-4aae-a079-42e06cd4be56 haproxy[2454]: [WARNING] 140/210508 (2454) : Former worker 2455 exited with code 0

```

## Do you have any idea what may have caused this?

I suspect the reload code flow is somehow interpreting the "-x" in the prefix of the string to be a "-x <unix_socket> get listening sockets from a unix socket" command line argument.

## Do you have an idea how to solve the issue?

Improve the command line parsing on reload to not mistake the local peer name value as another command line argument.

**Background on this issue:**
Historically the local peer name was/is limited to a short string. We track our instances using UUIDs which did not fit for the local peer name. We decided to sha the UUID then base64 encode it to get a unique, reproducible, local peer name. The base64 variant used was the URL encoding variant (for outside reasons). This means if we get the golden UUID formula, occasionally the base64 string will prefix with -x and the haproxy instance will be a time bomb waiting for a reload to fail. Oddly enough, the start command seems to not have a problem with the "-x" prefix.

I fully expect you to file this under "users do strange things" folder.

**Comment by TimWolla** [comment]
> I suspect the reload code flow is somehow interpreting the "-x" in the prefix of the string to be a "-x <unix_socket> get listening sockets from a unix socket" command line argument.

This is not the issue. The issue is within the `copy_argv` function, that skips certain command line parameters when building the command line for the re-exec:

https://github.com/haproxy/haproxy/blob/66163ec616bd3dfb06605bfc5a6e640ea7ada7d7/src/haproxy.c#L1341-L1368

Starting HAProxy as `./haproxy -d -f ./haproxy.cfg -W -L -xfoo` results in
```
"./haproxy"
"-d"
"-f"
"./haproxy.cfg"
"-W"
"-L"
"-sf"
"27528"
```
being passed along.

**Comment by TimWolla** [comment]
I was looking into preparing a patch, but I'm not entirely sure what the best solution would be going forward. There are multiple things to blame here:

1. `-L` accepts a local peer name starting with a hyphen, while other options, such as `-x` reject parameters starting with a hyphen.
2. The argument parser is too lenient with regard to flags, because it only ever tests prefixes. So `-What` is parsed as `-W`, while `-Wstrict-aliasing` is parsed as `-Ws`. IMO options should be rejected when there's any trailing characters that are not recognized (by using `strcmp` instead of testing each character individually).
3. `copy_argv` does not implement the full logic of the argument parser. So even if arguments are expected to not have trailing junk as discussed in (2) it would fail on a local peer name of literally `-x`.

I'd say (2) should be changed no matter what. The big question is then: Which of (1) or (3) should be done?

**Comment by johnsom** [comment]
I fully agree with your (2) opinion.

It seems like for the -L case, eating the next white space delineated "word" as the peer name would make sense. If someone makes a mistake of not including the peer name parameter, they will get a peer name of "-x" and not the "-x" flag set. Which, in this case also requires a parameter and would subsequently fail out with a missing parameter (one would hope).

**Comment by wlallemand** [comment]
> * `-L` accepts a local peer name starting with a hyphen, while other options, such as `-x` reject parameters starting with a hyphen.

That's probably the simplest workaround but also the better, using hyphens does not make much sense as arguments, it's ambiguous and not really useful, it will prevent new bugs and be more consistent. Also -L is supposed to replace the hostname, so starting an hostname with a hyphen is kind of weird.

For special cases where you need a file beginning with a hyphen, you can use ./-file or /path/-file, regarding configuration files the "--" option is also available.

The list of arguments is not limited for some options, for example with -sf/st, you need a new option beginning by a - or the end of the options to delimit the list of arguments.

> * The argument parser is too lenient with regard to flags, because it only ever tests prefixes. So `-What` is parsed as `-W`, while `-Wstrict-aliasing` is parsed as `-Ws`. IMO options should be rejected when there's any trailing characters that are not recognized (by using `strcmp` instead of testing each character individually).

I agree, we had the same problem with the configuration parser and we discovered a lot of people that were using invalid configurations because of this.

> 
> * `copy_argv` does not implement the full logic of the argument parser. So even if arguments are expected to not have trailing junk as discussed in (2) it would fail on a local peer name of literally `-x`.

If the arguments are validated in the first point it's not useful in my opinion. If we start reimplementing a logic for each arguments in this, we will have new bugs each time we add a new argument if it's forgotten. It must be simple.

**Comment by TimWolla** [comment]
> If the arguments are validated in the first point it's not useful in my opinion.

Yes, if (1) is changed then (3) does not need to be done.

**Comment by wlallemand** [comment]
After discussing with Willy we decided that's changing the behavior of the parser is not a good idea, and we don't want to break existing peers configuration.

I don't want to have the logic duplicated in copy_argv() which could introduce bugs for new options if we forgot to duplicate the code. But since we won't implement new options in versions <= 2.2, I'm going to duplicate the logic anyway.

For 2.3 we will change the way to declare options and the number of arguments they take so it could be more reliable. The hyphen could be forbidden only in the case of variable argument number this way (-sf/st).

**Comment by wlallemand** [comment]
I fixed the parser as well as related issues in the master:
0041741ef7253fcad2fc98f0b2a9968fb4af3574, a6b3249935cf8e4de53b9940419683ef82cb40b8, 4f71d304aad5199d9a0179deb34241b4a3665740, df6c5a8ffad7ec28ff6987dc6b1e26d68eb89c5e

So the general idea is to support hyphen at the beginning of an option argument if the number of argument for this option is know. Which means, every options but -sf/st, because these 2 options don't have a finite number of arguments, so they use the hyphen to guess the end of the list.

**Comment by johnsom** [comment]
Thank you for your work and support on this!

**Comment by wlallemand** [comment]
@johnsom The patches are now backported to the 2.1 branch, don't hesitate to test this branch before we issue a new 2.1 release.

**Comment by johnsom** [comment]
Confirmed, I used git commit 29e710624583d71e7eeae63a6ea7734006a1ea57 and cannot reproduce the problem on that build.

**Comment by wlallemand** [comment]
Thanks, 2.1.6 was released this morning, including this patchset.

**Comment by rohara** [comment]
It appears that the patches were backported to haproxy-1.8, yet I can still reproduce this problem in haproxy-1.8.27. And ideas why that might be?

**Comment by wlallemand** [comment]
Strange, it's working with the latest 1.9 but not with the 1.8, maybe we forgot something when backporting. I'm checking this.

**Comment by wlallemand** [comment]
I found out that by backporting df6c5a8, the support for fixing all the options was removed, probably by mistake.

**Comment by wlallemand** [comment]
@rohara a new fix is available in the 1.8 tree: https://git.haproxy.org/?p=haproxy-1.8.git;a=commit;h=58b3d8676bbef52bc76dd79ecfcf74582c34ec97

**Comment by rohara** [comment]
@wlallemand Thank you!
